### ISSUE TITLE
xnor on tensorflow

### ISSUE THREAD

**Issue by lychahaha** [jiecaoyu/XNOR-Net-PyTorch#16]
Thanks for your sharing. And i found some bug in cifar-10/main.py.
Line 23:in py3, keys() is a lazy method, it will raise exception if you change the keys when iterating. Adding a list() outside will be fine.
Line 123:batch size of test should be 128 instead of 100, as it correspond to the number in Line 78.
Line 165:+1 tab?

Actually, I implement your code on tensorflow. But it doesn't work. And i reduce the lr to 1e-4, it gets 80.9% acc.
I run your code and get a model of 86.17% acc. And I load the model in my code and get 86.2%acc. Amazing!!!
I find the problem that bn of the third binconv get a small computational error, which can be only detected by sum of the output of bn. and will be amplified by layers after that bn.
Maybe it's the problem of cuda or cudnn.T_T
My code is on https://github.com/ljhandlwt/xnor-net-tf

**Comment by jiecaoyu** [comment]
Hi @ljhandlwt , I think the Adam solver and some other implementation (BN) might be different between TF and PyTorch. Also, do you mind telling me the hyperparameters you use for your Adam solver?

**Comment by lychahaha** [comment]
Yes, of course.
learn rate=0.01(decay strategy is the same as yours)
beta1=0.9
beta2=0.999
eps=1e-8
weight decay=1e-5

**Comment by jiecaoyu** [comment]
@ljhandlwt , could you help check whether you have added [this line](https://github.com/jiecaoyu/XNOR-Net-PyTorch/blob/7f84f1bf6921bc99b05bbc869dddf2bd64989a40/CIFAR_10/util.py#L82) for your implementation?

**Comment by lychahaha** [comment]
@jiecaoyu 
Yes, [here](https://github.com/ljhandlwt/xnor-net-tf/blob/master/c11.py#L137).
But the results show that it's not necessary.

I found the worst acc is almost never lower than `0.1`, and it's because the gradient explosion cause the large parameters, making the prediction of some class always close to 1. So i reduce the learn rate to avoid the gradient explosion, and it works. Your `mul(n)` will enlarge the gradient and obviously it isn't the key of the problem.

Actually, i don't set a learn rate more than `1e-3` when i choose adam, since i never see some one set it more than `1e-3` in tensorflow. Is pytorch different?
And i notice a detail about weight decay. Adam in pytorch doesn't really calc the L2-loss and it apply it in the gradients directly, but tensorflow's will calc it and compute derivatives. 
As far as i know, adam has missing-precision problem. Maybe the implement of L2-loss is the key of the problem? I will check it soon.

**Comment by jiecaoyu** [comment]
@ljhandlwt 
Do you mean when I am modifying the gradients of the weights, I am actually modifying the gradients + L2-Norm gradients? I am not too sure about the implementation in PyTorch.

As for the LR, I am also using 1e-3 for the other networks. But only for NIN, it seems 0.01 works well.

**Comment by lychahaha** [comment]
@jiecaoyu 
Yes, [here](https://github.com/pytorch/pytorch/blob/master/torch/optim/adam.py#L89).
Adding L2-Norm(`y=1/2*x**2`) to the loss means adding the parameter itself to the gradient(`y'=x`).

Maybe i need to have a try of imagenet, however it 's a tiring work_(:з」∠)_.

**Comment by jiecaoyu** [comment]
@ljhandlwt Hi,
```python
grad = grad.add(group['weight_decay'], p.data)
```
means
```python
grad = grad + group['weight_decay'] * p.data
```
So I think the ```weight_decay``` can actually also be considered calculated?

In my implementation, the ```optimizer.step()``` function is called after ```bin_op.updateBinaryGradWeight()```. I think I am not modifying the gradients from L2-norm.

**Comment by lychahaha** [comment]
@jiecaoyu 
Oh, yes! Here's the point.

Your `bin_op.updateBinaryGradWeight()` modify gradient of softmax loss, without L2-norm.
It's not easy for tensorflow to do that. I will try to fix it soon, thanks!

**Comment by jiecaoyu** [comment]
@ljhandlwt yes, I got it! It is annoying haha.
